client/store/user.js

```diff
@@ -21,35 +21,40 @@ const removeUser = () => ({type: REMOVE_USER})
 /**
  * THUNK CREATORS
  */
-export const me = () => dispatch =>
-  axios
-    .get('/auth/me')
-    .then(res => dispatch(getUser(res.data || defaultUser)))
-    .catch(err => console.error(err))
+export const me = () => async dispatch => {
+  try {
+    const res = await axios.get('/auth/me')
+    dispatch(getUser(res.data || defaultUser))
+  } catch (err) {
+    console.error(err)
+  }
+}
+
+export const auth = (email, password, method) => async dispatch => {
+  let res
+  try {
+    res = await axios.post(`/auth/${method}`, {email, password})
+  } catch (authError) {
+    dispatch(getUser({error: authError}))
+  }
 
-export const auth = (email, password, method) => dispatch =>
-  axios
-    .post(`/auth/${method}`, {email, password})
-    .then(
-      res => {
-        dispatch(getUser(res.data))
-        history.push('/home')
-      },
-      authError => {
-        // rare example: a good use case for parallel (non-catch) error handler
-        dispatch(getUser({error: authError}))
-      }
-    )
-    .catch(dispatchOrHistoryErr => console.error(dispatchOrHistoryErr))
+  try {
+    dispatch(getUser(res.data))
+    history.push('/home')
+  } catch (dispatchOrHistoryErr) {
+    console.error(dispatchOrHistoryErr)
+  }
+}
```

Comment:
PPS it also occurs to me another way to deal with this would be to `return` inside the first `catch`. That might be less verbose overall.

Reply:
Hm, I'm torn here, going for the early return in the catch. But nothing about this feels right.

Reply:
I agree it is a nasty bit of the codebase. The bug that happened when it was simpler was especially awful, however, confused a lot of people in a very misleading way (making them think there was an error in their login logic, when in reality it would be a bug in a React component).

Probably there is a way to compose this thunk out of smaller async pieces, which would clean it up somewhat? Not sure offhand, and in any case I'd save that for a different PR.

Reply:
PS also this is an example of _one specific way_ that promises are actually _cleaner_ (sort of) than `async`/`await`. It is more difficult to branch on exception raising in sync-looking code.